src/mongo/shell.ts

```diff
@@ -5,120 +5,201 @@
+	private _onResult: EventEmitter<CommandResult> = new EventEmitter<CommandResult>();
 
-	public static create(execPath: string, connectionString: string, isEmulator: boolean): Promise<Shell> {
-		return new Promise((c, e) => {
+	public static create(execPath: string, execArgs: string, connectionString: string, isEmulator: boolean): Promise<Shell> {
+		return new Promise((resolve, reject) => {
+			let args: string[] = [];
 			try {
-				let args = ['--quiet', connectionString];
+				args = splitArguments(execArgs);
+				args = args.concat(['--quiet', connectionString]);
 				if (isEmulator) {
 					// Without this the connection will fail due to the self-signed DocDB certificate
 					args.push("--ssl");
 					args.push("--sslAllowInvalidCertificates");
 				}
 				const shellProcess = cp.spawn(execPath, args);
-				return c(new Shell(execPath, shellProcess));
+				return resolve(new Shell(execPath, args, shellProcess));
 			} catch (error) {
-				e(`Error while creating mongo shell with path '${execPath}': ${error}`);
+				reject(`Error while creating mongo shell with path '${execPath}' and arguments ${JSON.stringify(args)}: ${error}`);
 			}
 		});
 	}
 
-	constructor(private execPath: string, private mongoShell: cp.ChildProcess) {
-		this.initialize();
-	}
+	constructor(private _execPath: string, private _execArgs: string[], private _mongoShell: cp.ChildProcess) {
+		this._mongoShell.on('error', (error: unknown) => {
+			this.fireError(error);
+		});
+		this._mongoShell.on('exit', (exitCode: number, signal: string) => {
+			this._exited = true;
+
+			// One of exitCode/signal will always be non-null
+			let message = signal ? `Mongo shell exited with signal '${signal}.'` : `Mongo shell exited with code ${exitCode}.`;
+			this.fireError(message);
+		});
 
-	private initialize() {
-		const once = (ee: NodeJS.EventEmitter, name: string, fn: Function) => {
-			ee.once(name, fn);
-			this.disposables.push(toDisposable(() => ee.removeListener(name, fn)));
-		};
-
-		const on = (ee: NodeJS.EventEmitter, name: string, fn: Function) => {
-			ee.on(name, fn);
-			this.disposables.push(toDisposable(() => ee.removeListener(name, fn)));
-		};
-
-		once(this.mongoShell, 'error', result => this.onResult.fire(result));
-		once(this.mongoShell, 'exit', result => this.onResult.fire(result));
-
-		let buffers: string[] = [];
-		on(this.mongoShell.stdout, 'data', b => {
-			let data: string = b.toString();
-			const delimitter = `${this.executionId}${os.EOL}`;
-			if (data.endsWith(delimitter)) {
-				const result = buffers.join('') + data.substring(0, data.length - delimitter.length);
-				buffers = [];
-				this.onResult.fire({
-					exitCode: void 0,
-					result,
-					stderr: void 0
-				});
+		// Monitor STDOUT
+		this._mongoShell.stdout.on('data', (chunk: Buffer) => {
+			let data: string = chunk.toString();
```

Comment:
Perhaps it would be helpful to print `stdout` and `stderr` to an outputChannel (probably shouldn't be the main "Cosmos DB" one) as it happens? Rather than waiting until the end to show any progress?